Remove restriction on type parameters preceding consts w/ feature const-generics

JulianKnodt opened the pull request:
Removed the restriction on type parameters preceding const parameters when the feature const-generics is enabled.

Builds on #74676, which deals with unsorted generic parameters. This just lifts the check in lowering the AST to HIR that permits consts and types to be reordered with respect to each other. Lifetimes still must precede both

This change is not intended for min-const-generics, and is gated behind the `#![feature(const_generics)]`.

One thing is that it also permits type parameters without a default to come after consts, which I expected to not work, and was hoping to get more guidance on whether that should be permitted or how to prevent it otherwise.

I did not go through the RFC process for this pull request because there was prior work to get this feature added. In the previous PR that was cited, work was done to enable this change.

r? @lcnr

JulianKnodt committed:
Allow types to come after consts in AST validation

JulianKnodt committed:
Rm restriction on ord of default types w/ consts

JulianKnodt committed:
Added +1 test for only works w/ feat const gen

Added this test to ensure that reordering the parameters only works with the feature const
generics enabled.

Fixed nits

Also added another test to verify that intermixed lifetimes are forbidden

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @lcnr (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://rustc-dev-guide.rust-lang.org/contributing.html) for more information.

lcnr commented on src/test/ui/const-generics/defaults/right-order.rs:
```diff
@@ -0,0 +1,11 @@
+// run-pass
+// Verifies that having generic parameters after constants is permitted
```
```suggestion
// Verifies that having generic parameters after constants is permitted.
```

lcnr commented on src/test/ui/const-generics/defaults/needs-feature.rs:
```diff
@@ -0,0 +1,8 @@
+// Verifies that having generic parameters after constants is not permitted without the
+// const generics feature
```
```suggestion
// `const_generics` feature.
// FIXME(min_const_generics): Add the feature gate here.
```

lcnr reviewed (changes requested):
Nice to see that this already works!

The way `ParamKindOrd` works rn is somewhat unfortunate, I think either adding a field to `ParamKindOrd::Const(bool)` to specify if `const_generics` is active or adding a new `ParamKindOrd::MinConst` variant is the best way to fix this for now.

For example I would expect the following to still result in an incorrect help message.
```rust
struct Foo<const N: usize, 'a, T = u32>(&'a (), T);
```

lcnr commented:
cc @varkor @withoutboats @eddyb 

The goal of this PR is to remove the restriction on the param ordering between types and consts to allow for type defaults.

Should we only lift that restriction when using `feature(const_generics)` and keep disallowing this when using `feature(min_const_generics)`.

As you can see from this PR, keeping that restriction doesn't simplify the implementation. So keeping that restriction would mostly allow us keep forcing that parameter order in the future.

I personally am unsure how else we might want to implement default parameters here, so I would be fine with also allowing this as part of of the MVP.

JulianKnodt committed:
Test lifetimes after types after consts forbidden

Added more complex test and changed error message

JulianKnodt committed:
Blessed old test where error message had changed

Added minor fmt change to ast_validation

lzutao commented on src/librustc_ast_passes/ast_validation.rs:
```diff
@@ -735,6 +735,11 @@ fn validate_generic_param_order<'a>(
         let max_param = &mut max_param;
         match max_param {
+            Some(ParamKindOrd::Const)
+                if ParamKindOrd::Type == kind && sess.features_untracked().const_generics =>
+            {
+                ()
+            }
```
```suggestion
                if ParamKindOrd::Type == kind && sess.features_untracked().const_generics => {}
```

JulianKnodt committed:
Add ParamKindOrd::ConstUnordered variant

JulianKnodt commented:
Woops accidentally closed this

JulianKnodt committed:
Switched to unordered field in ParamKindOrd

Run fmt

JulianKnodt commented:
@bors try

bors commented:
@JulianKnodt: :key: Insufficient privileges: not in try users

JulianKnodt commented:
Hm it works locally with these changes

lcnr commented on src/test/ui/const-generics/argument_order.stderr:
```diff
@@ -1,39 +1,18 @@
-   |
-   = note: `#[warn(incomplete_features)]` on by default
-   = note: see issue #44580 <https://github.com/rust-lang/rust/issues/44580> for more information
+   |               -----------------^^-----^^-------------------- help: reorder the parameters: lifetimes, then consts and types: `<'a, 'b, T, U, const N: usize, const M: usize>`
```
You probably also have to overwrite the `Ord` impl for `ParamKindOrd` here.

This error messages should not reorder types and consts if `const_generics` is active

IOW this error message should mention 
```
help: reorder the parameters: lifetimes, then consts and types: `<'a, 'b, const N: usize, T, const M: usize, U>`
```

lcnr commented on src/librustc_ast/ast.rs:
```diff
@@ -313,15 +313,15 @@ pub type GenericBounds = Vec<GenericBound>;
     Lifetime,
     Type,
-    Const,
+    Const { unordered: bool },
```
can you also add a comment here mentioning that `min_const_generics` requires const params to come after types.

JulianKnodt committed:
Change Ord impl for ParamKindOrd

Updated tests and error msgs

Update stderr from test

Update w/ lcnr comments

Change some tests around, and also updated Ord implementation for ParamKindOrd

Update w/ nits from lcnr

bors commented:
:umbrella: The latest upstream changes (presumably #74877) made this pull request unmergeable. Please resolve the merge conflicts.

lcnr commented on src/librustc_ast/ast.rs:
```diff
@@ -309,19 +310,48 @@ pub type GenericBounds = Vec<GenericBound>;
     Lifetime,
     Type,
-    Const,
+    // unordered is only true if sess.has_features().const_generics
+    // is active. Specifically, if it's only `min_const_generics`, it will still require ordered
+    // types.
```
```suggestion
    // `unordered` is only `true` if `sess.has_features().const_generics`
    // is active. Specifically, if it's only `min_const_generics`, it will still require ordered
    // types.
```
nit: also what are "ordered types" here

JulianKnodt replied:
ah yea I can rephrase that to make sense my bad

lcnr commented on src/librustc_ast/ast.rs:
```diff
@@ -309,19 +310,48 @@ pub type GenericBounds = Vec<GenericBound>;
+impl PartialOrd for ParamKindOrd {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        use ParamKindOrd::*;
+        let ord = match self {
```
You can use `match (self, other)` here which is cleaner in my opinion.

Using `feature(or_patterns)` we can also write the following here:
```rust
(
    ParamKindOrd::Type | ParamKindOrd::Const { unordered: true },
    ParamKindOrd::Type | ParamKindOrd::Const { unordered: true }
) => true
```

lcnr commented on src/librustc_ast/ast.rs:
```diff
@@ -309,19 +310,48 @@ pub type GenericBounds = Vec<GenericBound>;
+}
+impl Ord for ParamKindOrd {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.partial_cmp(other).unwrap()
```
wouldn't it make more sense to implement `Ord` and use `Some(self.cmp(other))` in `PartialOrd`?

JulianKnodt replied:
ah yea I can do that

lcnr commented on src/test/ui/const-generics/defaults/complex-unord-param.rs:
```diff
@@ -0,0 +1,28 @@
+
+trait FixedOutputter {
+  fn out(&self) -> FixedOutput<'_, 10>;
+}
```
I don't think `FixedOutput` is used in the rest of the example...

It might make sense to move it into a different test, maybe `simple-defaults.rs` or sth :thinking:

lcnr commented on src/test/ui/const-generics/defaults/right-order.rs:
```diff
@@ -0,0 +1,11 @@
+#![feature(const_generics)]
+#![allow(incomplete_features)]
+
+struct A<const N: usize, T=u32>(T);
```
Imo it's probably better to unify this test with `src/test/ui/const-generics/defaults/needs-feature.rs` and use revisions here.

lcnr commented on src/test/ui/const-generics/defaults/type-after-const-requires-default.rs:
```diff
@@ -0,0 +1,10 @@
+// run-pass
```
afaik the name is somewhat misleading here?

I think we probably should move this out of `defaults`, considering that it doesn't use defaults

lcnr reviewed (approved):
A lot of small nits, once these are fixed r=me.

lcnr commented on src/librustc_ast/ast.rs:
```diff
@@ -314,35 +314,32 @@ pub type GenericBounds = Vec<GenericBound>;
-    // types.
+    // `unordered` is only `true` if `sess.has_features().const_generics`
+    // is active. Specifically, if it's only `min_const_generics`, it will still require
+    // ordering `const`s after `type`s.
```
```suggestion
    // ordering consts after types.
```
const and type aren't variables here :laughing:

JulianKnodt replied:
ah, I just assumed that anything that was a technical term was \` \`

lcnr commented on src/test/ui/const-generics/defaults/needs-feature.rs:
```diff
@@ -1,8 +1,15 @@
+// revisions: none const_gens
+
+#![cfg_attr(const_gens, feature(const_generics))]
+#![cfg_attr(const_gens, allow(incomplete_features))]
```
Can you add a revision for `min_const_generics` here?

lcnr commented on src/librustc_ast/ast.rs:
```diff
@@ -314,35 +314,32 @@ pub type GenericBounds = Vec<GenericBound>;
-        self.partial_cmp(other).unwrap()
+impl PartialOrd for ParamKindOrd {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(&other))
```
```suggestion
        Some(self.cmp(other))
```
afaict the ref is unnecessary here

lcnr commented on src/librustc_ast/ast.rs:
```diff
@@ -314,35 +314,32 @@ pub type GenericBounds = Vec<GenericBound>;
-            &Const { .. } if other == &Lifetime => Ordering::Greater,
-
-            &Const { unordered: false } if other == &Type => Ordering::Greater,
+        match (self, other) {
+            (&Lifetime, &Lifetime) => Ordering::Equal,
+            (&Lifetime, _) => Ordering::Less,
+            (_, &Lifetime) => Ordering::Greater,
+            (&Type | &Const { unordered: true }, &Type | &Const { unordered: true }) => {
+                Ordering::Equal
+            }
+            (&Type, &Const { unordered: false }) => Ordering::Less,
+            (&Const { unordered: false }, &Type) => Ordering::Greater,
```
nit: you can remove the `&` here due to match ergonomics.

If you don't want to rely on that, you can also match on `(*self, *other)` :thinking: 

I think in general that this match is slightly too complex for what it wants to do, maybe something like
```rust
let to_int = |v| match v {
    Lifetime => 0,
    Type | Const { unordered: true } => 1,
    Const { unordered: false } => 2,
};

to_int(self).cmp(to_int(other))
```

JulianKnodt committed:
Convert `Eq` impl to check Ord::Equal

lcnr commented:
@bors r+

bors commented:
:pushpin: Commit 64f643782245b25326c6e9a76bf78c798f1ae168 has been approved by `lcnr`

<!-- @bors r=lcnr 64f643782245b25326c6e9a76bf78c798f1ae168 -->
<!-- homu: {"type":"Approved","sha":"64f643782245b25326c6e9a76bf78c798f1ae168","approver":"lcnr"} -->

bors commented:
:hourglass: Testing commit 64f643782245b25326c6e9a76bf78c798f1ae168 with merge 04f5cefa734a20bd5edb13d43c0322720ac7174f...
<!-- homu: {"type":"BuildStarted","head_sha":"64f643782245b25326c6e9a76bf78c798f1ae168","merge_sha":"04f5cefa734a20bd5edb13d43c0322720ac7174f"} -->

bors commented:
:boom: Test timed out
<!-- homu: {"type":"TimedOut"} -->

lcnr commented:
@bors retry

bors commented:
:hourglass: Testing commit 64f643782245b25326c6e9a76bf78c798f1ae168 with merge 08324fe6f7ac24be4c8bfcab42b12ee447635c80...
<!-- homu: {"type":"BuildStarted","head_sha":"64f643782245b25326c6e9a76bf78c798f1ae168","merge_sha":"08324fe6f7ac24be4c8bfcab42b12ee447635c80"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/967593332), [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=35900)
Approved by: lcnr
Pushing 08324fe6f7ac24be4c8bfcab42b12ee447635c80 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"lcnr","base_ref":"master","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=35900","checks-actions":"https://github.com/rust-lang-ci/rust/runs/967593332"},"merge_sha":"08324fe6f7ac24be4c8bfcab42b12ee447635c80"} -->

varkor commented on src/test/ui/const-generics/defaults/needs-feature.min.stderr:
```diff
@@ -0,0 +1,8 @@
+  --> $DIR/needs-feature.rs:10:26
+   |
+LL | struct A<const N: usize, T=u32>(T);
+   |         -----------------^----- help: reorder the parameters: lifetimes, then consts, then types: `<T, const N: usize>`
```
This error message is wrong.

varkor commented on src/librustc_ast_passes/ast_validation.rs:
```diff
@@ -775,13 +775,13 @@ fn validate_generic_param_order<'a>(
+                if sess.features_untracked().const_generics {
+                    ", then consts and types"
+                } else if sess.features_untracked().min_const_generics {
+                    ", then consts, then types"
```
Shouldn't this be `", then types, then consts"`?

lcnr replied:
ups

lcnr replied:
@JulianKnodt do you want to open a PR for this, otherwise I will quickly fix that?

varkor reviewed (commented):
What's the reason we only allow consts and types to be reordered? If we're lifting the arbitrary restriction, we may as well lift it entirely.

varkor commented:
Sorry, I didn't have time to get to this before now.

lcnr commented:
> What's the reason we only allow consts and types to be reordered? If we're lifting the arbitrary restriction, we may as well lift it entirely.

mixing lt with types was forbidden before :laughing: I do feel like const and type parameters are more similar than lt params.

We also can't specify late bound lt params, which makes mixing them with other parameters somewhat undesirable imo.
(https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=06b9ccfc5d0c7ba4fd610635eda2179a)

Lifting the ordering restriction on lt bounds would also require an RFC, wouldn't it?

varkor commented:
> Lifting the ordering restriction on lt bounds would also require an RFC, wouldn't it?

Really, lifting the restriction even for normal const generics ought to require some discussion, as it wasn't part of the original RFC. I think if we're restricting at all, it makes sense to have `lifetimes, types, consts, type defaults, const defaults` as a strict order.
